### ISSUE TITLE
GSUB bug when decoding Castoro

### ISSUE THREAD

**Issue by RoelN** [Pomax/lib-font#123]
Twitter thread where this was first raised: https://twitter.com/TiroTypeworks/status/1356325050782150656

Font can be downloaded here: https://github.com/TiroTypeworks/Castoro (I used this [WOFF2](https://github.com/TiroTypeworks/Castoro/blob/master/fonts/Castoro/TTFWOFF2/Castoro-Regular.woff2))

Here's a reduced test case:

```JavaScript
import { Font } from "./lib-font.js";

const font = new Font("test");
font.src = `./fonts/broken/Castoro-Regular.woff2`;

font.onerror = (evt) => console.error(evt);
font.onload = (evt) => {
  let font = evt.detail.font;

  const GSUB = font.opentype.tables["GSUB"];
  const scripts = GSUB.getSupportedScripts();
  let allGlyphs = {};

  scripts.forEach((script) => {
    let langsys = GSUB.getSupportedLangSys(script);

    allGlyphs[script] = {};

    langsys.forEach((lang) => {
      let langSysTable = GSUB.getLangSysTable(script, lang);
      let features = GSUB.getFeatures(langSysTable);

      allGlyphs[script][lang] = {};

      features.forEach((feature) => {
        const lookupIDs = feature.lookupListIndices;
        allGlyphs[script][lang][feature.featureTag] = {};
        allGlyphs[script][lang][feature.featureTag]["lookups"] = [];

        lookupIDs.forEach((id) => {
          const lookup = GSUB.getLookup(id);

          // Castoro bugs here
          if (lookup.lookupType === 6) {
            lookup.subtableOffsets.forEach((_, i) => {
              console.log(`+++ Getting subtable ${i}, lookup ${id}, feature ${feature.featureTag}, script ${script} , lang ${lang} +++`);
              const subtable = lookup.getSubTable(i);
            });
          }
        });
      });
    });
  });
};
```

Running this yields:

```
+++ Getting subtable 0, lookup 6, feature ccmp, script latn , lang dflt +++
parser getUint16 2 Parser {
  name: 'GSUB',
  length: 6524,
  start: 4554,
  offset: 1972,
  data: DataView {
    byteLength: 6524,
    byteOffset: 0,
    buffer: ArrayBuffer {
      [Uint8Contents]: <00 01 00 00 00 0a 00 b4 01 66 00 01 6c 61 74 6e 00 08 00 22 00 05 43 41 54 20 00 3a 4d 4f 4c 20 00 6e 4e 4c 44 20 00 54 52 4f 4d 20 00 6e 54 52 4b 20 00 88 00 00 ff ff 00 09 00 00 00 01 00 02 00 03 00 04 00 05 00 0a 00 0b 00 0c 00 00 ff ff 00 0a 00 00 00 01 00 02 00 03 00 04 00 05 00 06 00 0a 00 0b ... 6424 more bytes>,
      byteLength: 6524
    }
  }
}
parser 4554 1972
Event {
  type: 'error',
  detail: RangeError: Offset is outside the bounds of the DataView
      at DataView.getUint16 (<anonymous>)
      at Parser.getValue (file:///Users/roel/code/lib-font/src/parser.js:60:29)
      at Parser.get (file:///Users/roel/code/lib-font/src/parser.js:38:25)
      at Parser.get Offset16 [as Offset16] (file:///Users/roel/code/lib-font/src/parser.js:130:17)
      at file:///Users/roel/code/lib-font/src/opentype/tables/advanced/shared/subtables/gsub.js:283:22
      at Array.map (<anonymous>)
      at new LookupType6 (file:///Users/roel/code/lib-font/src/opentype/tables/advanced/shared/subtables/gsub.js:283:9)
      at Object.buildSubtable (file:///Users/roel/code/lib-font/src/opentype/tables/advanced/shared/subtables/gsub.js:475:16)
      at LookupTable.getSubTable (file:///Users/roel/code/lib-font/src/opentype/tables/advanced/shared/lookup.js:53:20)
      at file:///Users/roel/code/lib-font/test.js:37:39,
  msg: 'Failed to load font at ./fonts/broken/Castoro-Regular.woff2'
}
```

**Comment by RoelN** [comment]
Possibly related to https://github.com/Pomax/lib-font/issues/105 ?

**Comment by Pomax** [comment]
I've slightly changed the test code to not single out lookup type 6:
```js
function testFont(font) {
  const { GSUB } = font.opentype.tables;
  const scripts = GSUB.getSupportedScripts();
  let allGlyphs = {};

  scripts.forEach((script) => {
    let langsys = GSUB.getSupportedLangSys(script);

    allGlyphs[script] = {};

    langsys.forEach((lang) => {
      let langSysTable = GSUB.getLangSysTable(script, lang);
      let features = GSUB.getFeatures(langSysTable);

      allGlyphs[script][lang] = {};

      features.forEach((feature) => {
        const lookupIDs = feature.lookupListIndices;
        allGlyphs[script][lang][feature.featureTag] = {};
        allGlyphs[script][lang][feature.featureTag]["lookups"] = [];

        lookupIDs.forEach((id) => {
          const lookup = GSUB.getLookup(id);

          lookup.subtableOffsets.forEach((_, i) => {
            console.log(
              `Getting subtable ${i}, lookup ${id}, feature ${feature.featureTag}, script ${script} , lang ${lang}`
            );
            const subtable = lookup.getSubTable(i);
          });
        });
      });
    });
  });
}
```

**Comment by Pomax** [comment]
This yields:
```
Getting subtable 0, lookup 14, feature c2sc, script latn , lang dflt
Getting subtable 0, lookup 15, feature c2sc, script latn , lang dflt
Getting subtable 0, lookup 17, feature case, script latn , lang dflt
Getting subtable 0, lookup 18, feature case, script latn , lang dflt
Getting subtable 0, lookup 4, feature ccmp, script latn , lang dflt
Getting subtable 0, lookup 5, feature ccmp, script latn , lang dflt
Getting subtable 0, lookup 6, feature ccmp, script latn , lang dflt
parser getUint16 2 Parser {
  name: 'GSUB',
  length: 6524,
  start: 4554,
  offset: 1972,
  data: DataView {
    byteLength: 6524,
    byteOffset: 0,
    buffer: ArrayBuffer {
      [Uint8Contents]: <00 01 00 00 00 0a 00 b4 01 66 00 01 6c 61 74 6e 00 08 00 22 00 05 43 41 54 20 00 3a 4d 4f 4c 20 00 6e 4e 4c 44 20 00 54 52 4f 4d 20 00 6e 54 52 4b 20 00 88 00 00 ff ff 00 09 00 00 00 01 00 02 00 03 00 04 00 05 00 0a 00 0b 00 0c 00 00 ff ff 00 0a 00 00 00 01 00 02 00 03 00 04 00 05 00 06 00 0a 00 0b ... 6424 more bytes>,
      byteLength: 6524
    }
  }
}
parser 4554 1972
RangeError: Offset is outside the bounds of the DataView
    at DataView.getUint16 (<anonymous>)
    at Parser.getValue (file:///C:/Users/Mike/Documents/Git/released/lib-font/src/parser.js:60:29)
    at Parser.get (file:///C:/Users/Mike/Documents/Git/released/lib-font/src/parser.js:38:25)
    at Parser.get Offset16 [as Offset16] (file:///C:/Users/Mike/Documents/Git/released/lib-font/src/parser.js:130:17) 
    at file:///C:/Users/Mike/Documents/Git/released/lib-font/src/opentype/tables/advanced/shared/subtables/lookuptypes/gsub-6.js:26:22
    at Array.map (<anonymous>)
    at new LookupType6 (file:///C:/Users/Mike/Documents/Git/released/lib-font/src/opentype/tables/advanced/shared/subtables/lookuptypes/gsub-6.js:26:9)
    at Object.buildSubtable (file:///C:/Users/Mike/Documents/Git/released/lib-font/src/opentype/tables/advanced/shared/subtables/gsub.js:24:16)
    at LookupTable.getSubTable (file:///C:/Users/Mike/Documents/Git/released/lib-font/src/opentype/tables/advanced/shared/lookup.js:53:20)
    at file:///C:/Users/Mike/Documents/Git/released/lib-font/testing/manual/custom/castoro-parsing.js:42:37
```

**Comment by Pomax** [comment]
Looking at `getSubTable()` and adding a try/catch:
```js
  getSubTable(index) {
    const builder = this.ctType === `GSUB` ? GSUBtables : GPOStables;
    this.parser.currentPosition = this.start + this.subtableOffsets[index];
    try {
      return builder.buildSubtable(this.lookupType, this.parser);
    } catch (e) {
      console.log(this.parser.currentPosition, this.start, this.subtableOffsets[index]);
      console.log(this.subtableOffsets);
      process.exit(1);
    }
  }
```
Yields:
```
Getting subtable 0, lookup 14, feature c2sc, script latn , lang dflt
Getting subtable 0, lookup 15, feature c2sc, script latn , lang dflt
Getting subtable 0, lookup 17, feature case, script latn , lang dflt
Getting subtable 0, lookup 18, feature case, script latn , lang dflt
Getting subtable 0, lookup 4, feature ccmp, script latn , lang dflt
Getting subtable 0, lookup 5, feature ccmp, script latn , lang dflt
Getting subtable 0, lookup 6, feature ccmp, script latn , lang dflt
parser getUint16 2 Parser {
  name: 'GSUB',
  length: 6524,
  start: 4554,
  offset: 1972,
  data: DataView {
    byteLength: 6524,
    byteOffset: 0,
    buffer: ArrayBuffer {
      [Uint8Contents]: <00 01 00 00 00 0a 00 b4 01 66 00 01 6c 61 74 6e 00 08 00 22 00 05 43 41 54 20 00 3a 4d 4f 4c 20 00 6e 4e 4c 44 20 00 54 52 4f 4d 20 00 6e 54 52 4b 20 00 88 00 00 ff ff 00 09 00 00 00 01 00 02 00 03 00 04 00 05 00 0a 00 0b 00 0c 00 00 ff ff 00 0a 00 00 00 01 00 02 00 03 00 04 00 05 00 06 00 0a 00 0b ... 6424 more bytes>,
      byteLength: 6524
    }
  }
}
parser 4554 1972
6526 4546 8
[ 8 ]
```

**Comment by Pomax** [comment]
Updating the GSUB lookup6 class:
```js
class LookupType6 extends LookupType {
  constructor(p) {
    super(p);

    console.log(`building GSUB lookup type 6 subtable, with substFormat ${this.substFormat}`);
```
Shows:
```
...
Getting subtable 0, lookup 6, feature ccmp, script latn , lang dflt
getSubTable(0): this.parser.currentPosition = 4546 + 8
building GSUB lookup type 6 subtable, with substFormat 2
```
So let's investigate the code specific to 6.2

**Comment by Pomax** [comment]
```js
...
    if (this.substFormat === 2) {
      this.coverageOffset = p.Offset16;
      this.backtrackClassDefOffset = p.Offset16;
      this.inputClassDefOffset = p.Offset16;
      this.lookaheadClassDefOffset = p.Offset16;
      this.chainSubClassSetCount = p.uint16;

      console.log(`Data found prior to binding this.chainSubClassSetOffsets:`, {
        coverageOffset: this.coverageOffset,
        backtrackClassDefOffset: this.backtrackClassDefOffset,
        inputClassDefOffset: this.inputClassDefOffset,
        lookaheadClassDefOffset: this.lookaheadClassDefOffset,
        chainSubClassSetCount: this.chainSubClassSetCount,
      });

      this.chainSubClassSetOffsets = [
        ...new Array(this.chainSubClassSetCount),
      ].map((_) => p.Offset16);
    }
...
```
 Yields:
```
Getting subtable 0, lookup 6, feature ccmp, script latn , lang dflt
getSubTable(0): this.parser.currentPosition = 4546 + 8
building GSUB lookup type 6 subtable, with substFormat 2
Data found prior to binding this.chainSubClassSetOffsets: {
  coverageOffset: 42,
  backtrackClassDefOffset: 130,
  inputClassDefOffset: 1556,
  lookaheadClassDefOffset: 15,
  chainSubClassSetCount: 1634
}
```

**Comment by Pomax** [comment]
```js
...
      console.log(`Trying to build this.chainSubClassSetOffsets...`)
      try {
        this.chainSubClassSetOffsets = [
          ...new Array(this.chainSubClassSetCount),
        ].map((_) => p.Offset16);
      } catch (e) {
        console.error(`yeah that failed.`);
        throw e;
      }
```
Shows
```
Getting subtable 0, lookup 6, feature ccmp, script latn , lang dflt
getSubTable(0): this.parser.currentPosition = 4546 + 8
building GSUB lookup type 6 subtable, with substFormat 2
Data found prior to binding this.chainSubClassSetOffsets: {
  coverageOffset: 42,
  backtrackClassDefOffset: 130,
  inputClassDefOffset: 1556,
  lookaheadClassDefOffset: 15,
  chainSubClassSetCount: 1634
}
Trying to build this.chainSubClassSetOffsets...
yeah that failed.
```

**Comment by Pomax** [comment]
Time to break out the documentation: https://docs.microsoft.com/en-us/typography/opentype/spec/gsub#lookuptype-6-chained-contexts-substitution-subtable

**Comment by Pomax** [comment]
In the subtable we see:
```js
class LookupType6 extends LookupType {
  constructor(p) {
    super(p);

    ...

    if (this.substFormat === 2) {
      this.coverageOffset = p.Offset16;
      this.backtrackClassDefOffset = p.Offset16;
      this.inputClassDefOffset = p.Offset16;
      this.lookaheadClassDefOffset = p.Offset16;
      this.chainSubClassSetCount = p.uint16;
```
But if we look at the super class, we see:
```js
class LookupType extends ParsedData {
  constructor(p) {
    super(p);
    this.substFormat = p.uint16;
    this.coverageOffset = p.Offset16;
  }
```
So there's our bug: we're double-parsing `this.coverageOffset`, once in the super class, and then _again_ in the specific class

**Comment by Pomax** [comment]
Removing the erroneous double parse from the `LookupType6 ` class:
```js
class LookupType6 extends LookupType {
  constructor(p) {
    super(p);

    ...

    if (this.substFormat === 2) {
      this.backtrackClassDefOffset = p.Offset16;
      this.inputClassDefOffset = p.Offset16;
      this.lookaheadClassDefOffset = p.Offset16;
      this.chainSubClassSetCount = p.uint16;
      this.chainSubClassSetOffsets = [
        ...new Array(this.chainSubClassSetCount),
      ].map((_) => p.Offset16);
    }

    ...
```
Now gives us:
```
Getting subtable 0, lookup 14, feature c2sc, script latn , lang dflt
Getting subtable 0, lookup 15, feature c2sc, script latn , lang dflt
Getting subtable 0, lookup 17, feature case, script latn , lang dflt
Getting subtable 0, lookup 18, feature case, script latn , lang dflt
Getting subtable 0, lookup 4, feature ccmp, script latn , lang dflt
Getting subtable 0, lookup 5, feature ccmp, script latn , lang dflt
Getting subtable 0, lookup 6, feature ccmp, script latn , lang dflt
Getting subtable 0, lookup 10, feature frac, script latn , lang dflt
Getting subtable 0, lookup 11, feature frac, script latn , lang dflt
Getting subtable 0, lookup 12, feature frac, script latn , lang dflt
Getting subtable 0, lookup 19, feature liga, script latn , lang dflt
Getting subtable 0, lookup 21, feature liga, script latn , lang dflt
Getting subtable 1, lookup 21, feature liga, script latn , lang dflt
Getting subtable 0, lookup 23, feature liga, script latn , lang dflt
Getting subtable 1, lookup 23, feature liga, script latn , lang dflt
Getting subtable 0, lookup 25, feature liga, script latn , lang dflt
Getting subtable 0, lookup 27, feature liga, script latn , lang dflt
Getting subtable 0, lookup 18, feature lnum, script latn , lang dflt
Getting subtable 0, lookup 16, feature smcp, script latn , lang dflt
Getting subtable 0, lookup 9, feature subs, script latn , lang dflt
Getting subtable 0, lookup 8, feature sups, script latn , lang dflt
Getting subtable 0, lookup 14, feature c2sc, script latn , lang CAT
Getting subtable 0, lookup 15, feature c2sc, script latn , lang CAT
Getting subtable 0, lookup 17, feature case, script latn , lang CAT
Getting subtable 0, lookup 18, feature case, script latn , lang CAT
Getting subtable 0, lookup 4, feature ccmp, script latn , lang CAT
Getting subtable 0, lookup 5, feature ccmp, script latn , lang CAT
Getting subtable 0, lookup 6, feature ccmp, script latn , lang CAT
Getting subtable 0, lookup 10, feature frac, script latn , lang CAT
Getting subtable 0, lookup 11, feature frac, script latn , lang CAT
Getting subtable 0, lookup 12, feature frac, script latn , lang CAT
Getting subtable 0, lookup 19, feature liga, script latn , lang CAT
Getting subtable 0, lookup 21, feature liga, script latn , lang CAT
Getting subtable 1, lookup 21, feature liga, script latn , lang CAT
Getting subtable 0, lookup 23, feature liga, script latn , lang CAT
Getting subtable 1, lookup 23, feature liga, script latn , lang CAT
Getting subtable 0, lookup 25, feature liga, script latn , lang CAT
Getting subtable 0, lookup 27, feature liga, script latn , lang CAT
Getting subtable 0, lookup 18, feature lnum, script latn , lang CAT
Getting subtable 0, lookup 0, feature locl, script latn , lang CAT
Getting subtable 0, lookup 16, feature smcp, script latn , lang CAT
Getting subtable 0, lookup 9, feature subs, script latn , lang CAT
Getting subtable 0, lookup 8, feature sups, script latn , lang CAT
Getting subtable 0, lookup 14, feature c2sc, script latn , lang MOL
Getting subtable 0, lookup 15, feature c2sc, script latn , lang MOL
Getting subtable 0, lookup 17, feature case, script latn , lang MOL
Getting subtable 0, lookup 18, feature case, script latn , lang MOL
Getting subtable 0, lookup 4, feature ccmp, script latn , lang MOL
Getting subtable 0, lookup 5, feature ccmp, script latn , lang MOL
Getting subtable 0, lookup 6, feature ccmp, script latn , lang MOL
Getting subtable 0, lookup 10, feature frac, script latn , lang MOL
Getting subtable 0, lookup 11, feature frac, script latn , lang MOL
Getting subtable 0, lookup 12, feature frac, script latn , lang MOL
Getting subtable 0, lookup 19, feature liga, script latn , lang MOL
Getting subtable 0, lookup 21, feature liga, script latn , lang MOL
Getting subtable 1, lookup 21, feature liga, script latn , lang MOL
Getting subtable 0, lookup 23, feature liga, script latn , lang MOL
Getting subtable 1, lookup 23, feature liga, script latn , lang MOL
Getting subtable 0, lookup 25, feature liga, script latn , lang MOL
Getting subtable 0, lookup 27, feature liga, script latn , lang MOL
Getting subtable 0, lookup 18, feature lnum, script latn , lang MOL
Getting subtable 0, lookup 2, feature locl, script latn , lang MOL
Getting subtable 0, lookup 16, feature smcp, script latn , lang MOL
Getting subtable 0, lookup 9, feature subs, script latn , lang MOL
Getting subtable 0, lookup 8, feature sups, script latn , lang MOL
Getting subtable 0, lookup 14, feature c2sc, script latn , lang NLD
Getting subtable 0, lookup 15, feature c2sc, script latn , lang NLD
Getting subtable 0, lookup 17, feature case, script latn , lang NLD
Getting subtable 0, lookup 18, feature case, script latn , lang NLD
Getting subtable 0, lookup 4, feature ccmp, script latn , lang NLD
Getting subtable 0, lookup 5, feature ccmp, script latn , lang NLD
Getting subtable 0, lookup 6, feature ccmp, script latn , lang NLD
Getting subtable 0, lookup 10, feature frac, script latn , lang NLD
Getting subtable 0, lookup 11, feature frac, script latn , lang NLD
Getting subtable 0, lookup 12, feature frac, script latn , lang NLD
Getting subtable 0, lookup 19, feature liga, script latn , lang NLD
Getting subtable 0, lookup 21, feature liga, script latn , lang NLD
Getting subtable 1, lookup 21, feature liga, script latn , lang NLD
Getting subtable 0, lookup 23, feature liga, script latn , lang NLD
Getting subtable 1, lookup 23, feature liga, script latn , lang NLD
Getting subtable 0, lookup 25, feature liga, script latn , lang NLD
Getting subtable 0, lookup 27, feature liga, script latn , lang NLD
Getting subtable 0, lookup 18, feature lnum, script latn , lang NLD
Getting subtable 0, lookup 1, feature locl, script latn , lang NLD
Getting subtable 0, lookup 16, feature smcp, script latn , lang NLD
Getting subtable 0, lookup 9, feature subs, script latn , lang NLD
Getting subtable 0, lookup 8, feature sups, script latn , lang NLD
Getting subtable 0, lookup 14, feature c2sc, script latn , lang ROM
Getting subtable 0, lookup 15, feature c2sc, script latn , lang ROM
Getting subtable 0, lookup 17, feature case, script latn , lang ROM
Getting subtable 0, lookup 18, feature case, script latn , lang ROM
Getting subtable 0, lookup 4, feature ccmp, script latn , lang ROM
Getting subtable 0, lookup 5, feature ccmp, script latn , lang ROM
Getting subtable 0, lookup 6, feature ccmp, script latn , lang ROM
Getting subtable 0, lookup 10, feature frac, script latn , lang ROM
Getting subtable 0, lookup 11, feature frac, script latn , lang ROM
Getting subtable 0, lookup 12, feature frac, script latn , lang ROM
Getting subtable 0, lookup 19, feature liga, script latn , lang ROM
Getting subtable 0, lookup 21, feature liga, script latn , lang ROM
Getting subtable 1, lookup 21, feature liga, script latn , lang ROM
Getting subtable 0, lookup 23, feature liga, script latn , lang ROM
Getting subtable 1, lookup 23, feature liga, script latn , lang ROM
Getting subtable 0, lookup 25, feature liga, script latn , lang ROM
Getting subtable 0, lookup 27, feature liga, script latn , lang ROM
Getting subtable 0, lookup 18, feature lnum, script latn , lang ROM
Getting subtable 0, lookup 2, feature locl, script latn , lang ROM
Getting subtable 0, lookup 16, feature smcp, script latn , lang ROM
Getting subtable 0, lookup 9, feature subs, script latn , lang ROM
Getting subtable 0, lookup 8, feature sups, script latn , lang ROM
Getting subtable 0, lookup 14, feature c2sc, script latn , lang TRK 
Getting subtable 0, lookup 15, feature c2sc, script latn , lang TRK
Getting subtable 0, lookup 17, feature case, script latn , lang TRK
Getting subtable 0, lookup 18, feature case, script latn , lang TRK
Getting subtable 0, lookup 4, feature ccmp, script latn , lang TRK
Getting subtable 0, lookup 5, feature ccmp, script latn , lang TRK
Getting subtable 0, lookup 6, feature ccmp, script latn , lang TRK
Getting subtable 0, lookup 10, feature frac, script latn , lang TRK
Getting subtable 0, lookup 11, feature frac, script latn , lang TRK
Getting subtable 0, lookup 12, feature frac, script latn , lang TRK
Getting subtable 0, lookup 19, feature liga, script latn , lang TRK
Getting subtable 0, lookup 21, feature liga, script latn , lang TRK
Getting subtable 1, lookup 21, feature liga, script latn , lang TRK
Getting subtable 0, lookup 23, feature liga, script latn , lang TRK
Getting subtable 1, lookup 23, feature liga, script latn , lang TRK
Getting subtable 0, lookup 25, feature liga, script latn , lang TRK
Getting subtable 0, lookup 27, feature liga, script latn , lang TRK
Getting subtable 0, lookup 18, feature lnum, script latn , lang TRK
Getting subtable 0, lookup 3, feature locl, script latn , lang TRK
Getting subtable 0, lookup 16, feature smcp, script latn , lang TRK
Getting subtable 0, lookup 9, feature subs, script latn , lang TRK
Getting subtable 0, lookup 8, feature sups, script latn , lang TRK
```

PR incoming

**Comment by Pomax** [comment]
PR #125 merged, v2.3.3 published to npm
